core/fedavgoptimizer.py

```diff
@@ -0,0 +1,144 @@
+		new_weights = result.results.get('weights')
+		self._update_weights(new_weights)
+		job_type = result.job.job_type
+		callback = callback_handler(job_type, self.LEVEL_2_JOB_CALLBACKS)
+		return callback(result)
+
+	def handle_new_info(self, payload):
+		"""
+		Level 1 Callback to handle the info from the network
+		Case 1: New weights
+		Case 2: Termination
+		Case 3: Init ???
+		"""
+		# event_type = payload['info_type']
+		# callback = callback_handler(event_type, self.LEVEL_2_INFO_CALLBACKS)
+		# NOTE: Hardcoded to handle new weights for now...
+		callback = callback_handler("new_weights", self.LEVEL_2_INFO_CALLBACKS)
+		return callback(payload)
+
+	def _update_weights(self, weights):
+		"""
+		Helper function to replace the weights in the optimizer's
+		currently stored DMLJob with these weights, therefore 
+		ensuring that any future DMLJobs will operate with the 
+		correct weights. Mutates, does not return anything.
+		Case 1: Initialization
+		Case 2: Training update
+		Case 3: Averaging update
+		"""
+		self.job.weights = weights
+
+	def _handle_new_weights(self, payload):
+		"""
+		Level 2 Callback to handle new weights info
+		Updates state and then returns appropriate DMLJob
+		"""
+		job = self.job
+		job.job_type = "average"
+		serialized_weights = payload["content"]
+		job.add_weights(serialized_weights)
+		return ActionableEventTypes.SCHEDULE_JOB.value, job
+	
+	def _done_training(self, payload):
+		"""
+		Level 2 Callback to handle a DMLResult of a training job
+		Should return a DMLJob of type Communication
+		"""
+		job = self.job
+		job.job_type = "communicate"
+		return ActionableEventTypes.SCHEDULE_JOB.value, job
+	
+	def _done_initializing(self, payload):
+		"""
+		Level 2 Callback to handle a DMLResult of an initialization job
+		Should return a DMLJob of type Train
+		"""
+		job = self.job
+		job.job_type = "train"
+		return ActionableEventTypes.SCHEDULE_JOB.value, job
+	
+	def _done_everything(self, payload):
+		"""
+		Level 2 Callback to terminate session
+		"""
+		return ActionableEventTypes.TERMINATE.value, None
+	
+	def _average_iter(self, payload):
+		"""
+		Level 2 Callback to handle a DMLResult of an averaging job
+		Should only return a DMLJob of type Train if averaging is done
+		"""
+		job = self.job
+		self.listen_iterations += 1
+		if self.listen_iterations > self.listen_bound:
+			job.job_type = "train"
+			return ActionableEventTypes.SCHEDULE_JOB.value, job
+		else:
+			return ActionableEventTypes.NOTHING.value, None
+
+	def kickoff(self):
+		"""
+		Kickoff method for the FedAvgOptimizer. 
+		"""
+		return ActionableEventTypes.SCHEDULE_JOB.value, self.job
+
+	def ask(self, event_type, payload):
+		"""
+		Parses an event dictionary into a level 1 callback.
+		If the callback is not defined, it does nothing.
+		Otherwise, calls the callback.
+		"""
+		callback = callback_handler(event_type, self.LEVEL_1_CALLBACKS)
+		return callback(payload)
+
+	def do_nothing(self, payload):
+		pass
```

Comment:
Since `ask()` returns a tuple, this should return `None, None` or something similar.